[EDNA-73] Extend Dashboard API

gromakovsky opened the pull request:
## Description

Problem: currently `edna-server` doesn't expose any way to create a new sub-experiment via its API.
Also there is no way to get experimental data file (it should be possible to download them) and metadata.

Solution:
1. Add a new endpoint to create sub-experiments.
2. Also add a similar one to compute 4PL parameters without saving a new sub-experiment.
3. Update `getMeasurements` to return IDs along with measurements
because sub-experiment creation endpoint takes measurement IDs as
argument.
4. Add endpoints to get metadata and raw experiment data file.

## Related issue(s)

https://issues.serokell.io/issue/EDNA-73

## :white_check_mark: Checklist for your Pull Request

<!--
Ideally a PR has all of the checkmarks set.

If something in this list is irrelevant to your PR, you should still set this
checkmark indicating that you are sure it is dealt with (be that by irrelevance).

If you don't set a checkmark (e. g. don't add a test for new functionality),
you must be able to justify that.
-->

#### Related changes (conditional)

- Tests
  - [x] If I added new functionality, I added tests covering it.
  - [x] If I fixed a bug, I added a regression test to prevent the bug from
        silently reappearing again.

- Documentation
  - [x] I checked whether I should update the docs and did so if necessary:
    - [README](../tree/master/README.md)

#### Stylistic guide (mandatory)

- [x] My commits comply with [the policy used in Serokell](https://www.notion.so/serokell/Commit-and-PR-policy-4cf98e1a910a415d86b5f2491d9af1af).
- [x] My code complies with the [style guide](../tree/master/docs/code-style.md).

gromakovsky committed:
Add sub-experiment creation to Dashboard API

Problem: currently edna-server doesn't expose any way to create
a new sub-experiment via its API.

Solution:
1. Add a new endpoint to create sub-experiments.
2. Update `getMeasurements` to return IDs along with measurements
because sub-experiment creation endpoint takes measurement IDs as
argument.

As a side effect, now `getMeasurements` fails for unknown
sub-experiments.

gromakovsky committed:
Add an endpoint to compute 4PL params without saving

Problem: during sub-experiment creation users want to know
how 4PL will be computed for the current state of changes.
Currently the only way to compute 4PL parameters is by creating a
new sub-experiment.

Solution: add a new endpoint that is like sub-experiment creation,
but doesn't save anything and returns only 4PL parameters.

gromakovsky committed:
Add experiment metadata getter to the server

Problem: it should be possible to view experiment's metadata, but
it's currently impossible to get it from the backend.

Solution: add a new endpoint that takes experiment ID and fetches
metadata and description for that experiment.

gromakovsky committed:
Add ability to download experiment files

Problem: users must be able to download files they previously uploaded,
but currently edna-server doesn't support it.

Solution: add a new endpoint that returns `octet-stream` content type
with raw bytes.
Specify `Content-Disposition` header, so that the file will be
suggested to be downloaded rather than opened in browser.

gromakovsky committed:
Add missing Dashboard tests

Problem: there are tests for the Dashboard service, but not all
functions and typical cases are covered. There is a TODO about that.

Solution:
1. Add tests for all recently added functions exposed from Service.
2. Add tests for old functions that weren't possible before
`newSubExperiment` addition.
3. Extract tests for getters into a separate spec to keep the main spec
not so big.

gromakovsky committed:
Improve error handling in deleteSubExperiment

Problem: currently `deleteSubExperiment` fails with `DECantDeletePrimary`
if it deletes nothing, but the reason may be that sub-experiment simply
does not exist.

Solution: check for existence and fail with DESubExperimentNotFound if
it's not found.

gromakovsky committed:
Return better stub 4PL parameters

This is a temporary change until we implement proper 4PL analysis
(should happen very soon).

sashasashasasha151 commented on backend/src/Edna/Dashboard/DB/Query.hs:
```diff
@@ -156,3 +205,12 @@ getRemovedMeasurements (SqlId subExperimentId) =
     pure removedMeasurement
+
+-- Helper
+getExperimentId :: SubExperimentId -> Edna Word32
+getExperimentId sqlId@(SqlId subExpId) = do
+  mRes <- runSelectReturningOne' $ select $ do
+    subExperiment <- all_ $ esSubExperiment ednaSchema
+    guard_ (seSubExperimentId subExperiment ==. val_ (SqlSerial subExpId))
+    pure (seExperimentId subExperiment)
+  justOrThrow (DESubExperimentNotFound sqlId) mRes
```
So, again, it looks more convenient to throw errors from Service part and here return Maybe Word32

sashasashasasha151 commented on backend/src/Edna/Dashboard/DB/Query.hs:
```diff
@@ -77,6 +86,24 @@ deleteSubExperiment (SqlId subExpId) =
+-- the same experiment ID).
+cloneSubExperiment :: SubExperimentId -> Text -> Params4PL -> Edna SubExperimentRec
+cloneSubExperiment sqlId newName result = do
+  experimentId <- getExperimentId sqlId
```
So, here and in `getMeasurements` I think `experimentId` should a function parameter.

gromakovsky replied:
Ok, I'll try, doesn't sound hard, but let me see.

sashasashasasha151 commented on backend/src/Edna/Dashboard/Service.hs:
```diff
@@ -64,10 +73,53 @@ setIsSuspiciousSubExperiment subExpId isSuspicious = do
 deleteSubExperiment :: SubExperimentId -> Edna NoContent
 deleteSubExperiment subExpId = NoContent <$ do
+  -- Check existence (not the optimal way, but it likely doesn't matter)
+  void $ getMeasurements subExpId
```
Not really understand what for you need it

gromakovsky replied:
Without this check we fail with `DECantDeletePrimary` for non-existing sub-experiments.

![image](https://user-images.githubusercontent.com/2556409/114175149-581b2f80-9942-11eb-9815-49f6bf6ad331.png)

sashasashasasha151 replied:
Them looks like you again need transact here

gromakovsky replied:
Ok, I'll add it.

sashasashasasha151 commented on backend/src/Edna/Analysis/FourPL.hs:
```diff
@@ -49,4 +53,4 @@ instance ToSchema Params4PL where
 -- TODO [EDNA-71] Implement!
 analyse4PL :: [(Double, Double)] -> IO Params4PL
-analyse4PL _ = pure (Params4PL 1 2 3 4)
+analyse4PL _ = pure (Params4PL 36404 1.14 33 -2552)
```
So, it will be useful but only for one experiment, but ok)

gromakovsky replied:
According to @pva701 
> Thx, mocks work for me. Will try

So I think it's ok indeed.

sashasashasasha151 commented on backend/src/Edna/Dashboard/Service.hs:
```diff
@@ -64,10 +73,53 @@ setIsSuspiciousSubExperiment subExpId isSuspicious = do
+  SubExperimentId -> NewSubExperimentReq -> Edna (WithId 'SubExperimentId SubExperimentResp)
+newSubExperiment subExpId req = do
+  (removed, newResult) <- analyseNewSubExperiment subExpId req
+  transact $ do
+    result <- subExperimentRecToResp <$>
+      Q.cloneSubExperiment subExpId (nserName req) newResult
+    result <$ insertRemovedMeasurements (wiId result) removed
```
Looks like transact is not needed

gromakovsky replied:
I think we should `cloneSubExperiment` and `insertRemovedMeasurements` atomically. If we insert sub-experiment and don't insert removed measurements (e. g. because the app crashes), we will get invalid data in DB.

From my experiments `transact` doesn't work like that, but it's a topic of another issue (EDNA-76). AFAIU it is supposed to give us atomicity, but doesn't give for some reason. Anyway, even if `transact` does nothing, it's a marker saying "these things should happen atomically".

sashasashasasha151 commented on backend/src/Edna/Dashboard/Web/API.hs:
```diff
@@ -71,6 +94,23 @@ data DashboardEndpoints route = DashboardEndpoints
+      :> Get '[JSON] ExperimentMetadata
+
+  , -- | Download experiment data file that stores experiment with given ID
+    deGetExperimentFile :: route
+      :- "experiment"
+      :> Summary "Download experiment data file that stores experiment with given ID"
+      :> Capture "experimentId" ExperimentId
+      :> "file"
+      :> Get '[OctetStream]
+        (Headers '[Header "Content-Disposition" Text] ExperimentFileBlob)
```
Is it possible somewhere to get file name and description? But as I see there is no such information in design

gromakovsky replied:
Yes, if you expand the table in Dashboard
![image](https://user-images.githubusercontent.com/2556409/114177111-b0533100-9944-11eb-82d5-1c6e73a81f67.png)

`Metadata` shows description and metadata from `.xlsx`, `Download` actually downloads `.xlsx` file.

sashasashasasha151 replied:
And file name?

gromakovsky replied:
I put file name into `Content-Disposition` header, so that when you click "Download" this file name will be used by default in the file save dialog.

sashasashasasha151 reviewed (changes requested):
Doesn't getSubExperiment should contain some flag about its primary? Also as I understand it is not possible to fetch all sub-experiments for experiment

gromakovsky commented:
> Doesn't getSubExperiment should contain some flag about its primary? 

This information is part of `ExperimentResp` and @pva701 seems to be ok with that.

> Also as I understand it is not possible to fetch all sub-experiments for experiment

Yes, @pva701 already wrote about that. See [here](https://serokell.slack.com/archives/C01F8AQSHGE/p1617727367281100?thread_ts=1617726546.279700&cid=C01F8AQSHGE). I was told that in practice the number of sub-experiments will most often be 2, 1 will be less often, then 3. More than 3 is a rare case. And also :are_you_serious: started arguing that making queries one-by-one is ok. So, taking all of that into account, I decided to zabeetb.

gromakovsky committed:
Deduplicate `insertSubExperiment` and `createSubExperiment`

Problem: `insertSubExperiment` and `createSubExperiment` functions
appeared to be very similar and mostly duplicate each other.

Solution: have only one function called `insertSubExperiment` that
all necessary data (including name) and returns whole SubExperimentRec.
